File: frontend/src/types/ingestion.ts
/**
 * Ingestion Types
 * 
 * Centralized type definitions for the HITL (Human-in-the-Loop) 
 * data ingestion flow including upload, processing, and mapping.
 */

// =============================================================================
// Matching Tier Types
// =============================================================================

/** Matching engine tier that produced the column mapping */
export type MatchingTier = 'hash' | 'fuzzy' | 'llm' | 'unmatched' | 'manual';

/** Status of a column mapping after processing */
export type MappingStatus = 'auto_mapped' | 'needs_review' | 'needs_attention';

/** Status of a data source */
export type DataSourceStatus = 'active' | 'processing' | 'archived' | 'error';

/** Status of a dry run record validation */
export type DryRunRecordStatus = 'valid' | 'warning' | 'error';

/** Overall status of a dry run preview */
export type DryRunOverallStatus = 'ready' | 'needs_review';

// =============================================================================
// Column Mapping Types
// =============================================================================

/**
 * Represents a single column mapping from source to target field.
 * Used in the mapping review step of the ingestion wizard.
 */
export interface ColumnMapping {
    /** Original column name from uploaded file */
    source_column: string;
    /** Canonical field name this column maps to, or null if unmapped */
    target_field: string | null;
    /** Confidence score of the match (0.0 to 1.0) */
    confidence: number;
    /** Which tier of the matching engine produced this match */
    tier: MatchingTier;
    /** Fuzzy matching score if applicable */
    fuzzy_score?: number;
    /** LLM reasoning if applicable */
    reasoning?: string;
    /** Sample values from this column */
    sample_data: unknown[];
    /** Whether this mapping requires user review */
    needs_review: boolean;
    /** Whether this column should be ignored during import */
    ignored: boolean;
    /** Current status of this mapping */
    status: MappingStatus;
}

/**
 * Statistics from the matching engine processing.
 */
export interface ProcessingStats {
    hash_matches: number;
    fuzzy_matches: number;
    llm_matches: number;
    unmatched: number;
    total_columns: number;
    hash_percent: number;
    fuzzy_percent: number;
    llm_percent: number;
    unmatched_percent: number;
}

/**
 * User confirmation for a single column mapping.
 */
export interface ColumnMappingConfirmation {
    source_column: string;
    target_field: string | null;
    ignored: boolean;
    user_corrected: boolean;
}

// =============================================================================
// Data Source Types
// =============================================================================

/**
 * Represents an uploaded data source (Excel/CSV file).
 * 
 * IMPORTANT: Files go through a HITL flow:
 * 1. Upload → Creates RawImport (raw_import_id)
 * 2. Process → Generates column mappings
 * 3. Confirm → Creates DataSource (id) and SchemaMapping
 * 
 * Files that haven't completed step 3 will have id: null
 */
export interface DataSource {
    /** DataSource ID - NULL if file hasn't been confirmed yet */
    id: string | null;
    /** RawImport ID - always present, identifies the uploaded file */
    raw_import_id: string;
    filename: string;
    uploaded_at: string;
    row_count: number;
    /** Current status in the HITL flow */
    ingestion_status: 'complete' | 'incomplete';
    status: DataSourceStatus;
    production_line_id?: string;
    factory_id?: string;
}

/**
 * Available canonical field for mapping dropdown.
 */
export interface AvailableField {
    field: string;
    description: string;
}

// =============================================================================
// API Request/Response Types
// =============================================================================

/**
 * Response from file upload endpoint.
 */
export interface UploadResponse {
    raw_import_id: string;
    filename: string;
    columns: number;
    rows: number;
    status: string;
    already_exists?: boolean;
}

/**
 * Response from file processing endpoint.
 */
export interface ProcessingResponse {
    raw_import_id: string;
    filename: string;
    columns: ColumnMapping[];
    stats: ProcessingStats;
    auto_mapped_count: number;
    needs_review_count: number;
    needs_attention_count: number;
}

/**
 * Request to confirm column mappings.
 */
export interface ConfirmMappingRequest {
    raw_import_id: string;
    mappings: ColumnMappingConfirmation[];
    /** Required: column containing date/time values */
    time_column: string;
    time_format?: string | null;
    data_source_id?: string;
    factory_id?: string;
    production_line_id?: string;
    learn_corrections?: boolean;
}

/**
 * Response from mapping confirmation.
 */
export interface ConfirmMappingResponse {
    schema_mapping_id: string;
    data_source_id: string;
    learned_aliases: number;
    message: string;
}

/**
 * File preview data with headers and sample rows.
 */
export interface FilePreview {
    raw_import_id: string;
    filename: string;
    headers: string[];
    sample_rows: unknown[][];
    total_rows: number;
    total_columns: number;
    status: string;
}

// =============================================================================
// Dry Run Preview Types
// =============================================================================

/**
 * A single record from dry run preview with validation status.
 */
export interface DryRunRecord {
    row_index?: number | null; // Optional: backend may not always provide this, or may return null
    raw_data: Record<string, unknown>;
    cleaned_data: Record<string, unknown>;
    status: DryRunRecordStatus;
    issues: string[];
}

/**
 * Response from dry run preview endpoint.
 */
export interface DryRunResponse {
    raw_import_id: string;
    total_rows: number;
    preview_records: DryRunRecord[];
    column_mapping_used: Record<string, string>;
    overall_status: DryRunOverallStatus;
}

File: backend/app/schemas/ingestion.py
# Copyright (c) 2026 Aaron Guo. All rights reserved.
# Use of this source code is governed by the proprietary license
# found in the LICENSE file in the root directory of this source tree.

# backend\app\schemas\ingestion.py
"""
Pydantic schemas for Data Ingestion (HITL Flow).

This module defines schemas for:
- File processing and column mapping
- User confirmation of mappings
- Dry-run preview responses

The HITL (Human-in-the-Loop) flow:
1. Upload file → Create RawImport record
2. Process file → Run through Waterfall Matching Engine
3. Confirm mapping → Save SchemaMapping and learn corrections
"""

from datetime import datetime
from typing import Any

from pydantic import BaseModel, Field, model_validator

from app.enums import MatchTier


class ColumnMappingResult(BaseModel):
    """
    API response for a single column mapping result.

    Represents the engine's best guess for mapping a source column
    to a canonical target field, with confidence scoring.
    """

    source_column: str = Field(
        ..., description="Original column name from uploaded file"
    )
    target_field: str | None = Field(
        None, description="Canonical field name (e.g., 'actual_qty', 'production_date')"
    )
    confidence: float = Field(
        ..., ge=0, le=1, description="Confidence score (0.0 to 1.0)"
    )
    tier: MatchTier = Field(
        ...,
        description="Matching tier used: 'hash', 'fuzzy', 'llm', 'manual', 'unmatched'",
    )
    fuzzy_score: float | None = Field(
        None, description="RapidFuzz similarity score (0-100) if tier='fuzzy'"
    )
    reasoning: str | None = Field(
        None, description="Explanation from matching engine (especially for LLM tier)"
    )
    sample_data: list[Any] = Field(
        default_factory=list,
        description="Sample values from this column (first 5 non-null)",
    )
    needs_review: bool = Field(
        False, description="True if confidence is medium (0.6-0.9)"
    )
    ignored: bool = Field(
        False, description="True if column should be skipped during import"
    )
    status: str = Field(
        ...,
        description="Status: 'auto_mapped', 'needs_review', 'needs_attention', 'ignored'",
    )


class ProcessingResponse(BaseModel):
    """
    Response from processing a file through the waterfall matching engine.

    Contains all column mapping results with statistics.
    """

    raw_import_id: str = Field(..., description="UUID of the RawImport record")
    filename: str = Field(..., description="Original filename")
    columns: list[ColumnMappingResult] = Field(
        ..., description="Mapping results for each column"
    )
    stats: dict[str, Any] = Field(
        default_factory=dict,
        description="Engine statistics (tier counts, processing time)",
    )
    auto_mapped_count: int = Field(
        ..., description="Number of columns auto-mapped with high confidence"
    )
    needs_review_count: int = Field(
        ..., description="Number of columns needing user review"
    )
    needs_attention_count: int = Field(
        ..., description="Number of columns that couldn't be mapped"
    )


# =============================================================================
# Confirmation Schemas
# =============================================================================


class ColumnMappingConfirmation(BaseModel):
    """
    User's confirmed mapping for a single column.

    Used when submitting final mappings after review.
    """

    source_column: str = Field(..., description="Original column name from file")
    target_field: str | None = Field(
        None, description="Canonical field to map to (None if ignored)"
    )
    ignored: bool = Field(False, description="True if column should be skipped")
    user_corrected: bool = Field(
        False, description="True if user changed the AI suggestion"
    )


class ConfirmMappingRequest(BaseModel):
    """
    Request to confirm column mappings after user review.

    This is step 3 of the HITL flow. Saves the mapping and optionally
    learns from user corrections for future matching.
    """

    raw_import_id: str = Field(..., description="UUID of the RawImport to confirm")
    mappings: list[ColumnMappingConfirmation] = Field(
        ..., description="List of confirmed column mappings"
    )
    time_column: str = Field(
        ..., description="REQUIRED: Source column name for time/date data"
    )
    time_format: str | None = Field(
        None, description="Date format (e.g., 'YYYY-MM-DD', 'DD/MM/YYYY')"
    )
    data_source_id: str | None = Field(
        None, description="UUID of existing DataSource to associate with"
    )
    production_line_id: str | None = Field(
        None, description="UUID of ProductionLine (required if no data_source_id)"
    )
    factory_id: str | None = Field(
        None, description="UUID of Factory for scoped alias learning"
    )
    learn_corrections: bool = Field(
        True,
        description="Whether to save corrections to alias database for future matching",
    )

    @model_validator(mode="after")
    def check_unique_target_fields(self) -> "ConfirmMappingRequest":
        """Ensure no two columns map to the same target field."""
        target_fields = [
            m.target_field for m in self.mappings if m.target_field and not m.ignored
        ]
        duplicates = [f for f in set(target_fields) if target_fields.count(f) > 1]
        if duplicates:
            raise ValueError(f"Duplicate target field mappings: {duplicates}")
        return self


class ConfirmMappingResponse(BaseModel):
    """Response after confirming mapping."""

    schema_mapping_id: str = Field(
        ..., description="UUID of created/updated SchemaMapping"
    )
    data_source_id: str = Field(..., description="UUID of associated DataSource")
    learned_aliases: int = Field(
        ..., description="Number of new aliases learned from corrections"
    )
    message: str = Field(..., description="Human-readable success message")


# =============================================================================
# Utility Schemas
# =============================================================================


class AvailableField(BaseModel):
    """Available canonical field for UI dropdown."""

    field: str = Field(..., description="Canonical field name (e.g., 'actual_qty')")
    description: str = Field(..., description="Human-readable description")


class DryRunRow(BaseModel):
    """Single row in dry-run preview showing before/after."""

    row_number: int = Field(..., description="Row number in original file (1-indexed)")
    original: dict[str, Any] = Field(
        ..., description="Original data as key-value pairs"
    )
    transformed: dict[str, Any] = Field(
        ..., description="Transformed data using confirmed mapping"
    )
    validation_errors: list[str] = Field(
        default_factory=list, description="Validation errors for this row"
    )
    validation_warnings: list[str] = Field(
        default_factory=list, description="Validation warnings for this row"
    )


class DryRunResponse(BaseModel):
    """Response for dry-run preview endpoint."""

    raw_import_id: str = Field(..., description="UUID of the RawImport")
    filename: str = Field(..., description="Original filename")
    total_rows: int = Field(..., description="Total rows in file")
    preview_rows: int = Field(..., description="Number of rows in preview")
    rows: list[DryRunRow] = Field(
        ..., description="Preview rows with before/after data"
    )
    column_mapping: dict[str, str] = Field(
        ..., description="Active column mapping used for transformation"
    )
    error_count: int = Field(..., description="Total validation errors across all rows")
    warning_count: int = Field(
        ..., description="Total validation warnings across all rows"
    )


class RawImportUploadResponse(BaseModel):
    """Response from file upload endpoint in the ingestion flow."""

    raw_import_id: str = Field(..., description="UUID of created RawImport record")
    filename: str = Field(..., description="Original filename")
    columns: int = Field(..., description="Number of columns detected")
    rows: int = Field(..., description="Number of rows detected")
    already_exists: bool = Field(
        False, description="True if file was already uploaded (deduplication)"
    )


class UploadListItem(BaseModel):
    """Single item in uploads list."""

    id: str
    original_filename: str
    file_type: str | None
    file_size_bytes: int
    row_count: int | None
    status: str
    data_source_id: str | None
    created_at: datetime | None
    factory_id: str | None
    production_line_id: str | None


class UploadListResponse(BaseModel):
    """Response for uploads list endpoint."""

    files: list[UploadListItem]
    total: int
    offset: int
    limit: int


class PreviewResponse(BaseModel):
    """
    Response for previewing uploaded file data before processing.
    """
    data: list[list[Any]] = Field(..., description="First N rows of data for preview")
    columns: list[str] = Field(..., description="Column headers detected")
    preview_rows: int = Field(..., description="Number of rows included in this preview")
    total_rows: int = Field(..., description="Total rows in the file")
    total_columns: int = Field(..., description="Total columns in the file")
    filename: str = Field(..., description="Name of the uploaded file")
    status: str = Field(..., description="Current status of the raw import")
